### PR TITLE
SE: Support binary OR

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Peach finding

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 1]
```diff
@@ -68,7 +66,17 @@ private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBin
...
-            if (left.HasConstraint<BoolConstraint>() && right.HasConstraint<BoolConstraint>())
+            if (left is null && right is null)
+            {
+                return null;
+            }
+            else if (left is null || right is null)
+            {
+                return kind == BinaryOperatorKind.Or && (left ?? right).HasConstraint(BoolConstraint.True)
+                    ? BoolConstraint.True
+                    : null;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:77]: The same is true for BinaryOperatorKind.And (with BoolConstraint.False).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 2]
```diff
@@ -359,6 +359,19 @@ void StringEmpty7(string path)
         {
             var s = string.IsNullOrEmpty(path) ? new string[] { } : path.Split('/');
         }
+
+        void BinaryOr(int i)
+        {
+            object o = null;
+            if (o==null | i == 0)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:366]: Add a test case with the null check on the right hand side and short circuiting `if (i == 0 || o == null)`.

Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:366]:
```suggestion
            if (o == null | i == 0)
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 3]
```diff
@@ -186,6 +186,26 @@ public void Binary_BoolOperands_Or()
 else
     Tag(""False | False"");
 
+if (isTrue | arg)
+    Tag(""isTrue | arg True "");
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:190]:
```suggestion
    Tag(""isTrue | arg True"");
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 4]
```diff
@@ -195,7 +215,18 @@ public void Binary_BoolOperands_Or()
     Tag(""False || True"");
 else
     Tag(""False || True Unreachable"");";
-            SETestContext.CreateCS(code).Validator.ValidateTagOrder("True | True", "False | True", "False | False", "True || True", "False || True");
+            SETestContext.CreateCS(code, ", bool arg").Validator.ValidateTagOrder(
+                "True | True",
+                "False | True",
+                "False | False",
+                "isTrue | arg True ",
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:222]:
```suggestion
                "isTrue | arg True",
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Add "And" support and some formatting issues.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:366]: || is not a binary operation but a different CFG shape

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:77]: I'll do another PR

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Pull request merged**